### PR TITLE
feat(capture-rs): implement replay capture in capture-rs

### PR DESCRIPTION
**Pull request by frankh**
## Problem

capture is fragmented

django capture for recordings is flaky

lets move everything to rust

## Changes

👉 _Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review._

## How did you test this code?

added unit tests

### CODE REVIEW

File: rust/capture/src/v0_endpoint.rs [hunk 1]
```diff
@@ -130,30 +118,130 @@ pub async fn event(
 
     let billing_limited = state
         .billing
-        .is_limited(context.token.as_str(), QuotaResource::Events)
+        .is_limited(context.token.as_str(), quota_resource)
         .await;
 
     if billing_limited {
         report_dropped_events("over_quota", events.len() as u64);
```
Comment (oliverb123) [inline rust/capture/src/v0_endpoint.rs:125]: I _think_ this should be moved to line 171?

**Review by oliverb123 — COMMENTED** [review]
This is a much smaller change than I expected tbh, all looks good to me aside from 1 question about where a metric is reported

File: rust/capture/src/v0_endpoint.rs (hunk 1 above)
Comment (oliverb123) [reply rust/capture/src/v0_endpoint.rs:125]: The reporting specifically

File: rust/capture/src/v0_endpoint.rs [hunk 2]
```diff
@@ -226,3 +326,53 @@ pub async fn process_events<'a>(
...
+                "$session_id": session_id,
+                "$window_id": window_id,
+                "$snapshot_source": events[0].properties.get("$snapshot_source").unwrap_or(&Value::String(String::from("web"))),
+                "$snapshot_items": snapshot_items,
+            }
+        }).to_string(),
+        now: context.now.clone(),
+        sent_at: context.sent_at,
+        token: context.token.clone(),
+        session_id: Some(session_id.as_str().unwrap().to_string()),
```
Comment (oliverb123) [inline rust/capture/src/v0_endpoint.rs:374]: I'd really rather not `unwrap()` here. Could we change line 348 onwards to something like:
```rust
    let session_id = events[0]
        .properties
        .get("$session_id")
        .ok_or(CaptureError::MissingSessionId)?
        .as_str()
        .ok_or(CaptureError::InvalidSessionId)? // Adding this error type ofc
        .to_string();
```

File: rust/capture/tests/recordings.rs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        "$session_id": session_id,
+        "properties": {
+            "$window_id": window_id,
+            "$snapshot_data": [],
+        }
+    });
+    let res = server.capture_recording(event.to_string()).await;
+    assert_eq!(StatusCode::BAD_REQUEST, res.status());
+    Ok(())
+}
```
Comment (oliverb123) [inline rust/capture/tests/recordings.rs:69]: Here's a (currently failing) test for the issue above, anyone could panic (and therefore DOS) replay capture by sending badly crafted session ID's due to that unwrap:
```rust
#[tokio::test]
async fn it_rejects_bad_session_id() -> Result<()> {
    setup_tracing();
    let token = random_string("token", 16);
    let distinct_id = random_string("id", 16);
    let window_id = random_string("id", 16);

    let main_topic = EphemeralTopic::new().await;
    let server = ServerHandle::for_recordings(&main_topic).await;

    let event = json!({
        "token": token,
        "event": "testing",
        "distinct_id": distinct_id,
        "$session_id": {"should_not_be": "an object"},
        "properties": {
            "$session_id": {"should_not_be": "an object"},
            "$window_id": window_id,
            "$snapshot_data": [],
        }
    });
    let res = server.capture_recording(event.to_string()).await;
    assert_eq!(StatusCode::BAD_REQUEST, res.status());
    Ok(())
}
```

**Review by oliverb123 — CHANGES_REQUESTED** [review]
Only one note, a bit of a security/stability issue. As a general rule, you should never `unwrap()` in any "real" code path, only tests, without being absolutely certain the thing being unwrapped will be what you expect (and usually the thing that makes you certain of that also lets you avoid `unwrap()`)

File: rust/capture/src/v0_endpoint.rs (hunk 2 above)
Comment (frankh) [reply rust/capture/src/v0_endpoint.rs]: good catch, fixed

**Review by oliverb123 — APPROVED** [review]

**Pull request merged**